cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java

```diff
@@ -0,0 +1,125 @@
+import javax.annotation.Nullable;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+
+/**
+ * {@link co.cask.http.HttpHandler} for exposing metadata of a runtime.
+ */
+@Path("v1/runtime")
+public class RuntimeHandler extends AbstractHttpHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(RuntimeHandler.class);
+  private static final Gson GSON = new Gson();
+  private static final Type MAP_STRING_CONSUME_REQUEST_TYPE = new TypeToken<Map<String,
+    MonitorConsumeRequest>>() { }.getType();
+
+  private final CConfiguration cConf;
+  private final MessageFetcher messageFetcher;
+
+  public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher) {
+    this.cConf = cConf;
+    this.messageFetcher = messageFetcher;
+  }
+
+  /**
+   * Gets list of topics along with offsets and limit as request and returns list of messages
+   */
+  @POST
+  @Path("/metadata")
+  public void metadata(FullHttpRequest request, HttpResponder responder) throws Exception {
+    String requestBody = request.content().toString(StandardCharsets.UTF_8);
+
+    Map<String, MonitorConsumeRequest> consumeRequests = GSON.fromJson(requestBody, MAP_STRING_CONSUME_REQUEST_TYPE);
+
+    ChunkResponder chunkResponder = responder.sendChunkStart(
+      HttpResponseStatus.OK, new DefaultHttpHeaders().set(HttpHeaderNames.CONTENT_TYPE,
+                                                          "application/json; charset=utf-8"));
+
+    ByteBuf buffer = Unpooled.buffer();
+    JsonWriter jsonWriter = new JsonWriter(new OutputStreamWriter(new ByteBufOutputStream(buffer),
+                                                                  StandardCharsets.UTF_8));
+    jsonWriter.beginObject();
+
+    for (Map.Entry<String, MonitorConsumeRequest> entry : consumeRequests.entrySet()) {
+      jsonWriter.name(entry.getKey());
+      jsonWriter.beginArray();
+
+      writeMessages(jsonWriter, buffer, chunkResponder, cConf.get(entry.getKey()), entry.getValue().getLimit(),
+                    entry.getValue().getMessageId());
+
+      jsonWriter.endArray();
+    }
+
+    jsonWriter.endObject();
+    jsonWriter.close();
+
+    if (buffer.isReadable()) {
+      chunkResponder.sendChunk(buffer.copy());
+    }
+
+    Closeables.closeQuietly(chunkResponder);
+  }
+
+  private void writeMessages(JsonWriter jsonWriter, ByteBuf buffer, ChunkResponder chunkResponder, String topic,
+                             int limit, @Nullable String fromMessage) throws TopicNotFoundException, IOException {
+    try (CloseableIterator<Message> iter = messageFetcher.fetch(NamespaceId.SYSTEM.getNamespace(), topic, limit,
+                                                                fromMessage)) {
+      while (iter.hasNext()) {
+        Message message = iter.next();
+        GSON.toJson(new MonitorMessage(message.getId(), message.getPayloadAsString(StandardCharsets.UTF_8)),
+                    MonitorMessage.class, jsonWriter);
+        if (buffer.isReadable()) {
```

Comment:
You are effectively writing out every message in separate chunk, since you are not using `BufferedWriter` for the `JsonWriter`.